include/ideep/operators/matmul.hpp

```diff
@@ -818,23 +819,42 @@ struct matmul_forward : public dnnl::matmul,
     // introduces *an extra reorder* afterwards. Here we keep the weight format
     // untouched thanks to optimizations for both plain and transposed formats
     // in DNNL.
-    IDEEP_ENFORCE(weights.get_data_type() == data_type::f32 ||
-                  weights.get_data_type() == data_type::bf16 ||
-                  weights.get_data_type() == data_type::f16,
-                  "Incorrect data type in weights");
-    dst_data_type = src.get_data_type() == data_type::bf16 ?
-                    data_type::bf16 :
-                    ((src.get_data_type() == data_type::f16) ?
-                      data_type::f16 : data_type::f32);
-    src_desc = src.get_desc().to_type(dst_data_type);
+    IDEEP_ENFORCE(
+        weights.get_data_type() == data_type::f32 ||
+            weights.get_data_type() == data_type::bf16 ||
+            weights.get_data_type() == data_type::f16,
+        "Incorrect data type in weights");
 
 #ifdef __aarch64__
+
+    dst_data_type = src.get_data_type() == data_type::bf16
+        ? ((dst.get_data_type() == data_type::f32) ? data_type::f32
+                                                   : data_type::bf16)
+        : ((src.get_data_type() == data_type::f16) ? data_type::f16
+                                                   : data_type::f32);
+    src_desc = (src.get_data_type() == data_type::bf16 &&
+                dst_data_type == data_type::f32)
+        ? src.get_desc()
+        : src.get_desc().to_type(dst_data_type);
     // for aarch64 ACL backend with fixed format kernels, the weights are
     // always in blocked layout, so, set the descriptor to tag::any for the backend
     // to decide the format
-    weights_desc = tensor::desc(weights.get_dims(), dst_data_type, tag::any);
+    wei_data_type = (src.get_data_type() == data_type::bf16 &&
+                   dst_data_type == data_type::f32) ? data_type::bf16 : dst_data_type;
+    weights_desc = tensor::desc(weights.get_dims(), wei_data_type, tag::any);
 #else
 
+    dst_data_type = src.get_data_type() == data_type::bf16
+        ? ((dst.get_data_type() == data_type::f32) ? data_type::f32
+                                                   : data_type::bf16)
+        : ((src.get_data_type() == data_type::f16) ? data_type::f16
+                                                   : data_type::f32);
+
+    src_desc = (src.get_data_type() == data_type::bf16 &&
+                dst_data_type == data_type::f32)
+        ? src.get_desc()
+        : src.get_desc().to_type(dst_data_type);
```

Comment:
If the logic is the same from both paths, suggest to avoid dup here.

Reply:
I have a question here, why this logic will be applied outside `__aarch64` macro? Although it will get the same result when src and dst are bf16, it is unnecessary.

Reply:
Would this incur different semantics for the API when it works on different CPU archs? Normally, we shouldn't design an API like that way. Is it possible to add an extra arg to specify the behavior you added (dst type does not follow src)?

Reply:
In this case I think this logic was generally used before. I added a diff condition very specific for aarch64

 >Would this incur different semantics for the API when it works on different CPU archs?
 
https://github.com/intel/ideep/blob/3215f5fe2d53fe7ba865a036dac4fc511d3e569e/include/ideep/operators/matmul.hpp#L825 
I shifted the logic in `ideep_pytorch` which was generically for all arch to when cpu_arch != aarch64. 
Please let me know you thoughts.

Reply:
That means the semantics are different between aarch64 and x86, right? I found that the arg `dst_type` is not used. Can we use it to override the `dst_data_type` to make sure the consistent behavior between aarch64 and x86? Of course, we have to redefine the default value for `dst_type` (it is now fp32 which doesn't seem right for the default behavior).

Reply:
I see, I do understand. 
I have updated the dst_type default value to `undef`. The logic which uses `dst_type` in this context here:  https://github.com/intel/ideep/blob/c54a3ede72ae05b483ee81353e94227c3057424d/include/ideep/operators/matmul.hpp#L884  should make sense now. Please let me know your thoughts.

Reply:
Then, can we specify the `dst_type` from the caller side? You don't have to use different logic to initialize `dst_data_type` on aarch64 and x64?